### PR TITLE
Projections follow up

### PR DESCRIPTION
**Pull request by mdumandag**
This is a follow up PR after merging the Projections PR. This is
a combination of lots of small things.

- Add documentation to properties of `MapEntry` so that they are
displayed
- Move map#project so that we maintain alphabetical order
- Add some corner case tests for projections and aggregations
- Used asserCountEqual on projection tests so that the tests will
be more durable, even if we add more items to map
- Add missing API documentation for projections
- Fix API documentation of projections around the return value documentations
- Add unit tests for the invalid projection inputs
- Make the projections code snippet simpler
- Add a code sample for projections

### CODE REVIEW

File: hazelcast/projection.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def get_class_id(self):
 
 
 class _MultiAttributeProjection(_AbstractProjection):
-    def __init__(self, *attribute_paths):
+    def __init__(self, attribute_paths):
```
Comment (yuce) [inline hazelcast/projection.py:51]: How about adding types to new classes and functions?

File: tests/integration/backward_compatible/proxy/map_test.py [hunk 2]
```diff
@@ -796,6 +796,14 @@ def setUp(self):
     def tearDown(self):
         self.map.destroy()
 
+    def test_aggregate_with_none_aggregator(self):
+        with self.assertRaises(AssertionError):
+            self.map.aggregate(None)
```
Comment (yuce) [inline tests/integration/backward_compatible/proxy/map_test.py:801]: Can be simplified as: `self.assertRaises(AssertionError, lambda: self.map.aggregate(None))` I think most of other assertRaises uses can also be put into a single line.

**Review by yuce — CHANGES_REQUESTED** [review]
Looks good, but I think using `attribute` instead of `single_attribute` and `multi_attribute` would make our API simpler and more Pythonic.

File: tests/integration/backward_compatible/proxy/map_test.py (hunk 2 above)
Comment (mdumandag) [reply tests/integration/backward_compatible/proxy/map_test.py:801]: I personally prefer this style, and try to use it like this across the codebase. Advantages:

- You can easily put breakpoints
- You can put multiple lines without defining a closure

And IMHO, this one is more readable

File: docs/using_python_client_with_hazelcast_imdg.rst [hunk 3]
```diff
@@ -2166,31 +2166,30 @@ See the following example.
...
-    employee_multi_attribute = employees.project(multi_attribute("Age", "Height"))
-    # Prints:
-    # Employee 1 age and height: [21, 170] Employee 2 age and height: [23, 180]
-    print("Employee 1 age and height: %s Employee 2 age and height: %s" % (employee_multi_attribute[0], employee_multi_attribute[1]))
+    filtered_ages = employees.project(single_attribute("age"), greater("age", 23))
+
+    # Prints: "Ages of the filtered employees are [25, 40]"
+    print("Ages of the filtered employees are %s" % filtered_ages)
+
+    attributes = employees.project(multi_attribute("age", "height"))
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply docs/using_python_client_with_hazelcast_imdg.rst:2189]: But they are two distinct things with distinct return types. One returns `list[any]`, the other `list[list[any]]` when used with project method. 

Also, I find such codes like the below a bit weird in the implementation. IMHO, a separation is better than this

```python
def attribute(*attrs):
    if len(attrs) == 1:
        return _SingleAttribute(attrs[0])
    return _MultiAttribute(attrs)
```

File: hazelcast/projection.py (hunk 1 above)
Comment (mdumandag) [reply hazelcast/projection.py:53]: I have added type hints to the projections module. I didn't type-hint the code that is coming from the superclass (IdentifiedDataSerializable), should I add type hints to those methods too (like get_class_id)? 

Also, I question. We have docstrings that define types for public methods, should we define type hints for them too? I have added type hints to them but I can remove them if you want

**Review by yuce — APPROVED** [review]
